Zend/zend_builtin_functions.c

```diff
@@ -434,16 +432,18 @@ static int validate_constant_array(HashTable *ht) /* {{{ */
 			if (Z_TYPE_P(val) == IS_ARRAY) {
 				if (Z_REFCOUNTED_P(val)) {
 					if (Z_IS_RECURSIVE_P(val)) {
-						zend_error(E_WARNING, "Constants cannot be recursive arrays");
+						zend_argument_value_error(argument_number, "cannot be a recursive array");
 						ret = 0;
 						break;
-					} else if (!validate_constant_array(Z_ARRVAL_P(val))) {
+					} else if (!validate_constant_array_argument(Z_ARRVAL_P(val), argument_number)) {
 						ret = 0;
 						break;
 					}
 				}
 			} else if (Z_TYPE_P(val) != IS_STRING && Z_TYPE_P(val) != IS_RESOURCE) {
-				zend_error(E_WARNING, "Constants may only evaluate to scalar values, arrays or resources");
+				zend_argument_type_error(2, "must be of type bool|int|float|string|array|resource|null, %s given",
```

Comment:
I think it would be more helpful to just say what *isn't* allowed at this point -- which is just objects, right?

Reply:
Yes, only objects. I'll change the message according to this!